--- a/linux/drivers/media/dvb/frontends/stb0899_algo.c.orig	2011-05-22 19:42:56.159506943 +0200
+++ a/linux/drivers/media/dvb/frontends/stb0899_algo.c	2011-05-22 19:49:10.179507169 +0200
@@ -342,38 +342,42 @@
 	u8 reg;
 
 	msleep(1);
-	internal->status = NODATA;
+	reg = stb0899_read_reg(state, STB0899_VSTATUS);
+	lock = STB0899_GETFIELD(VSTATUS_LOCKEDVIT, reg);
+	if ( !lock ) {
+
+		internal->status = NODATA;
+		msleep(1);
+
+		/* RESET FEC    */
+		reg = stb0899_read_reg(state, STB0899_TSTRES);
+		STB0899_SETFIELD_VAL(FRESACS, reg, 1);
+		stb0899_write_reg(state, STB0899_TSTRES, reg);
+		msleep(1);
+		reg = stb0899_read_reg(state, STB0899_TSTRES);
+		STB0899_SETFIELD_VAL(FRESACS, reg, 0);
+		stb0899_write_reg(state, STB0899_TSTRES, reg);
+		if (params->srate <= 2000000)
+			dataTime = 2000;
+		else if (params->srate <= 5000000)
+			dataTime = 1500;
+		else if (params->srate <= 15000000)
+			dataTime = 1000;
+		else
+			dataTime = 500;
+
+		stb0899_write_reg(state, STB0899_DSTATUS2, 0x00); /* force search loop	*/
+		while (1) {
+			/* WARNING! VIT LOCKED has to be tested before VIT_END_LOOOP	*/
+			reg = stb0899_read_reg(state, STB0899_VSTATUS);
+			lock = STB0899_GETFIELD(VSTATUS_LOCKEDVIT, reg);
+			loop = STB0899_GETFIELD(VSTATUS_END_LOOPVIT, reg);
+			if (lock || (loop && index) || (index > dataTime))
+				break;
+			index++;
+			msleep(1);
+		}
 
-	/* RESET FEC	*/
-	reg = stb0899_read_reg(state, STB0899_TSTRES);
-	STB0899_SETFIELD_VAL(FRESACS, reg, 1);
-	stb0899_write_reg(state, STB0899_TSTRES, reg);
-	msleep(1);
-	reg = stb0899_read_reg(state, STB0899_TSTRES);
-	STB0899_SETFIELD_VAL(FRESACS, reg, 0);
-	stb0899_write_reg(state, STB0899_TSTRES, reg);
-	msleep(1);
-
-	if (params->srate <= 2000000)
-		dataTime = 2000;
-	else if (params->srate <= 5000000)
-		dataTime = 1500;
-	else if (params->srate <= 15000000)
-		dataTime = 1000;
-	else
-		dataTime = 500;
-
-	stb0899_write_reg(state, STB0899_DSTATUS2, 0x00); /* force search loop	*/
-	while (1) {
-		msleep(1); 		// Alex: added 1 mSec
-		/* WARNING! VIT LOCKED has to be tested before VIT_END_LOOOP	*/
-		reg = stb0899_read_reg(state, STB0899_VSTATUS);
-		lock = STB0899_GETFIELD(VSTATUS_LOCKEDVIT, reg);
-		loop = STB0899_GETFIELD(VSTATUS_END_LOOPVIT, reg);
-
-		if (lock || loop || (index > dataTime))
-			break;
-		index++;
 	}
 
 	if (lock) {	/* DATA LOCK indicator	*/
