diff -r 4987f8a59d8a linux/drivers/media/dvb/frontends/ds3000.c
--- a/linux/drivers/media/dvb/frontends/ds3000.c	Sat Jun 18 00:53:46 2011 +0300
+++ b/linux/drivers/media/dvb/frontends/ds3000.c	Sun Jul 10 12:04:59 2011 +0300
@@ -1136,7 +1136,10 @@
 			ds3000_writereg(state,
 				ds3000_dvbs2_init_tab[i],
 				ds3000_dvbs2_init_tab[i + 1]);
-		ds3000_writereg(state, 0xfe, 0x98);
+		if (c->symbol_rate >= 30000000)
+			ds3000_writereg(state, 0xfe, 0x54);
+		else
+			ds3000_writereg(state, 0xfe, 0x98);
 		break;
 	default:
 		return 1;
