diff -pru a/linux/drivers/media/dvb/dvb-usb//az6027.c b/linux/drivers/media/dvb/dvb-usb//az6027.c
--- a/linux/drivers/media/dvb/dvb-usb//az6027.c	2011-07-03 13:40:40.000000000 +0000
+++ b/linux/drivers/media/dvb/dvb-usb//az6027.c	2011-07-03 13:46:37.000000000 +0000
@@ -1089,6 +1089,7 @@ static struct usb_device_id az6027_usb_t
 	{ USB_DEVICE(USB_VID_TERRATEC,  USB_PID_TERRATEC_DVBS2CI_V2) },
 	{ USB_DEVICE(USB_VID_TECHNISAT, USB_PID_TECHNISAT_USB2_HDCI_V1) },
 	{ USB_DEVICE(USB_VID_TECHNISAT, USB_PID_TECHNISAT_USB2_HDCI_V2) },
+	{ USB_DEVICE(USB_VID_ELGATO,	USB_PID_ELGATO_EYETV_SAT) },
 	{ },
 };
 
diff -pru a/linux/drivers/media/dvb/dvb-usb//dvb-usb-ids.h b/linux/drivers/media/dvb/dvb-usb//dvb-usb-ids.h
--- a/linux/drivers/media/dvb/dvb-usb//dvb-usb-ids.h	2011-07-03 13:40:40.000000000 +0000
+++ b/linux/drivers/media/dvb/dvb-usb//dvb-usb-ids.h	2011-07-03 13:44:04.000000000 +0000
@@ -295,6 +295,7 @@
 #define USB_PID_MSI_DIGI_VOX_MINI_III                   0x8807
 #define USB_PID_SONY_PLAYTV				0x0003
 #define USB_PID_MYGICA_D689				0xd811
+#define USB_PID_ELGATO_EYETV_SAT			0x002a
 #define USB_PID_ELGATO_EYETV_DIVERSITY			0x0011
 #define USB_PID_ELGATO_EYETV_DTT			0x0021
 #define USB_PID_ELGATO_EYETV_DTT_Dlx			0x0020
